### PR TITLE
Serverside RUI, but it doesn't stink ( hopefully )

### PR DESCRIPTION
**Pull request by F1F7Y**
Allows servers to use rui presets on clients, clients dont need to update to play

# Polls
```cpp
void function NSCreatePollOnPlayer( entity player, string header, array<string> options, float duration );
int function NSGetPlayerResponse( entity player );
```
poll-image

# Large message
```cpp
void function NSSendLargeMessageToPlayer( entity player, string title, string description, float duration, string image );
```
large

# Info
```cpp
void function NSSendInfoMessageToPlayer( entity player, string text );
```
info

# PopUp
```cpp
void function NSSendPopUpMessageToPlayer( entity player, string text );
```
popup

# Announcement
```cpp
void function NSSendAnnouncementMessageToPlayer( entity player, string title, string description, vector color, int priority, int style );
```
announcement

# Status
```cpp
void function NSCreateStatusMessageOnPlayer( entity player, string title, string description, string id );
void function NSEditStatusMessageOnPlayer( entity player, string title, string description, string id  );
void function NSDeleteStatusMessageOnPlayer( entity player, string id  );
```
2status

### CODE REVIEW

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+}
+
+void function DeleteStatusMessage( string id )
+{
+	if ( id in client.statusMessageList )
+	{
+		var rui = client.statusMessageList[ id ]
+		RuiSetGameTime( rui, "startFadeOutTime", Time() )
+		
+		// Remove it form table
```
Comment (GeckoEidechse) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:315]: Typo? ^^

```suggestion
		// Remove it from table
```

**Review by uniboi — APPROVED** [review]
[Odd.RuiDemo.zip](https://github.com/R2Northstar/NorthstarMods/files/9143832/Odd.RuiDemo.zip)

Code looks fine and tested all features hud elements, looks good to me!

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut (hunk 1 above)
Comment (F1F7Y) [reply Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:315]: :)

**Review by GeckoEidechse — APPROVED** [review]
Worked fine in testing

(many thanks to @uniboi for the demo mod <3)

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+{
+	while( true )
+	{
+		while( client.infoMessageQueue.len() == 0 )
+			WaitFrame()
+
+		var rui = CreatePermanentCockpitRui( $"ui/death_hint_mp.rpak" )
+		RuiSetString( rui, "hintText", client.infoMessageQueue[0].description )
+		RuiSetGameTime( rui, "startTime", Time() )
+		RuiSetFloat3( rui, "bgColor", <0,0,0> )
```
Comment (RoyalBlue1) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:437]:
```suggestion
		RuiSetFloat3( rui, "bgColor", < 0, 0, 0 > )
```

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+	client.poll.pollActive = true
+	
+	
+	for( int i = 0; i < client.poll.options.len() + 1; i++ )
+	{
+		var rui = CreateCockpitRui( $"ui/cockpit_console_text_top_left.rpak" )
+		// This makes it fade and me no likey >:(
+		//RuiSetInt( rui, "maxLines", 1 )
+		//RuiSetInt( rui, "lineNum", i )
+		RuiSetFloat2( rui, "msgPos", <0, 0.4 + i * 0.025, 0> )
```
Comment (RoyalBlue1) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:394]:
```suggestion
		RuiSetFloat2( rui, "msgPos", < 0, 0.4 + i * 0.025, 0 > )
```

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+	// Cap at 4 messages at a time
+	if( client.statusMessageList.len() == STATUS_MESSAGES_MAX )
+		return
+
+	var rui = CreatePermanentCockpitRui( $"ui/at_wave_intro.rpak" )
+	RuiSetInt( rui, "listPos", client.statusMessageList.len() )
+	RuiSetGameTime( rui, "startFadeInTime", Time() )
+	RuiSetString( rui, "titleText", client.temp.title )
+	RuiSetString( rui, "itemText", client.temp.description )
+	RuiSetFloat2( rui, "offset", <0,-250,0> )
```
Comment (RoyalBlue1) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:354]:
```suggestion
	RuiSetFloat2( rui, "offset", < 0, -250, 0 > )
```

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+	table<entity,int> playerPollResponses
+} server
+#endif // SERVER
+
+
+#if CLIENT
+global function MessageUtils_ClientInit
+
+vector ColorSelected = <0.9, 0.8, 0.5>
+vector ColorBase = <0.9, 0.5, 0.1>
```
Comment (RoyalBlue1) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:27]:
```suggestion
vector ColorSelected = < 0.9, 0.8, 0.5 >
vector ColorBase = < 0.9, 0.5, 0.1 >
```

**Review by RoyalBlue1 — COMMENTED** [review]
The spaces between the numbers and the bracket are not always in Respawn code but Gecko wants them to be there. The ones between the numbers themselves are always there.

Pls fix this as a batch commit

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+	
+	array<tempMessage> largeMessageQueue
+	array<tempMessage> popupMessageQueue
+	array<tempMessage> announcementQueue
+	array<tempMessage> infoMessageQueue
+	
+	// table<id,rui>
+	table<string,var> statusMessageList
+	
+} client
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:64]:
```suggestion
	table<string,var> statusMessageList
} client
```
hate new lines

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+		wait 0.6
+		
+		RuiDestroyIfAlive( rui )
+		
+		int i = 0
+		foreach( _id, _rui in client.statusMessageList )
+		{
+			RuiSetInt( _rui, "listPos", i )
+
+			i++
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:328]:
```suggestion
			RuiSetInt( _rui, "listPos", i )
			i++
```
looks ugly idk

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+
+void function ShowPollMessage_Threaded()
+{
+	if( client.poll.pollActive )
+		return
+	
+	client.poll.pollActive = true
+	
+	
+	for( int i = 0; i < client.poll.options.len() + 1; i++ )
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:388]:
```suggestion
	client.poll.pollActive = true
	
	for( int i = 0; i < client.poll.options.len() + 1; i++ )
```
do you like new lines or sumn, one is enough

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+	{
+		var rui = CreateCockpitRui( $"ui/cockpit_console_text_top_left.rpak" )
+		// This makes it fade and me no likey >:(
+		//RuiSetInt( rui, "maxLines", 1 )
+		//RuiSetInt( rui, "lineNum", i )
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:393]:
```suggestion
```
dont need those comments do we? if you want those as a guide on how shit works maybe make it comprehensive n stuff

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+		client.poll.ruis.append( rui )
+	}
+	
+	client.poll.options.clear()
+	
+	float endTime = Time() + client.poll.duration
+	while( endTime > Time() && client.poll.pollActive )
+		WaitFrame()
+	
+
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:419]:
```suggestion
	
```
newline my beloved

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+		while( client.infoMessageQueue.len() == 0 )
+			WaitFrame()
+
+		var rui = CreatePermanentCockpitRui( $"ui/death_hint_mp.rpak" )
+		RuiSetString( rui, "hintText", client.infoMessageQueue[0].description )
+		RuiSetGameTime( rui, "startTime", Time() )
+		RuiSetFloat3( rui, "bgColor", <0,0,0> )
+		RuiSetFloat( rui, "bgAlpha", 0.5 )
+
+		wait 7
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:440]: should make this a constant, convar or able to set within the command that calls this function. just waiting however long you think is nice might fuck you over in the future

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+		AnnouncementData announcement = Announcement_Create( client.announcementQueue[0].title )
+		Announcement_SetSubText( announcement, client.announcementQueue[0].description )
+		Announcement_SetTitleColor( announcement, client.announcementQueue[0].color )
+		Announcement_SetPurge( announcement, true )
+		Announcement_SetPriority( announcement, client.announcementQueue[0].priority )
+		Announcement_SetSoundAlias( announcement, SFX_HUD_ANNOUNCE_QUICK )
+		Announcement_SetStyle( announcement, client.announcementQueue[0].style )
+		AnnouncementFromClass( GetLocalViewPlayer(), announcement )
+
+		wait 5
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:463]: should make this a constant, convar or able to set within the command that calls this function. just waiting however long you think is nice might fuck you over in the future

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+		while( client.popupMessageQueue.len() == 0 )
+			WaitFrame()
+
+		var rui = CreateCockpitRui( $"ui/killdeath_info.rpak" )
+		RuiSetGameTime( rui, "startTime", Time() )
+		RuiSetFloat( rui, "duration", 20 ) // It has a weird end animation
+		RuiSetString( rui, "messageText", client.popupMessageQueue[0].description )
+		RuiSetBool( rui, "isBigText", true )
+
+		wait 2.4
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:503]: should make this a constant, convar or able to set within the command that calls this function. just waiting however long you think is nice might fuck you over in the future

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 14]
```diff
@@ -0,0 +1,510 @@
...
+
+		var rui = CreatePermanentCockpitRui( $"ui/death_hint_mp.rpak" )
+		RuiSetString( rui, "hintText", client.infoMessageQueue[0].description )
+		RuiSetGameTime( rui, "startTime", Time() )
+		RuiSetFloat3( rui, "bgColor", <0,0,0> )
+		RuiSetFloat( rui, "bgAlpha", 0.5 )
+
+		wait 7
+
+		client.infoMessageQueue.remove(0)
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:442]:
```suggestion
		client.infoMessageQueue.remove( 0 )
```
we do the shitty spaces everywhere so why not here?

File: Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut [hunk 15]
```diff
@@ -0,0 +1,510 @@
...
+		RuiSetBool( rui, "isBigText", true )
+
+		wait 2.4
+
+		client.popupMessageQueue.remove(0)
+		RuiDestroyIfAlive( rui )
+	}
+}
+
+#endif // CLIENT
```
Comment (ScureX) [inline Northstar.Custom/mod/scripts/vscripts/sh_message_utils.gnut:510]:
```suggestion
}
#endif // CLIENT
```

we didnt add a new line after the \#if SERVER so why add one before the \#endif

**Review by ScureX — CHANGES_REQUESTED** [review]

**Comment by F1F7Y** [comment]
Me when firefox cached the page and I didn't see the reviews :)

**Comment by uniboi** [comment]
> The spaces between the numbers and the bracket are not always in Respawn code but Gecko wants them to be there. The ones between the numbers themselves are always there.
> 
> Pls fix this as a batch commit

this worsens readability and I've seen respawn formatting vectors as `<x, y, z>` more often than `< x, y, z >`

**Comment by ASpoonPlaysGames** [comment]
> > The spaces between the numbers and the bracket are not always in Respawn code but Gecko wants them to be there. The ones between the numbers themselves are always there.
> > Pls fix this as a batch commit
> 
> this worsens readability and I've seen respawn formatting vectors as `<x, y, z>` more often than `< x, y, z >`

Counterpoint: `<x, y, z>` creates a compile error if x is negative so i prefer the other option

**Comment by uniboi** [comment]
ok I changed my mind I just used `<(x), y z>` for no reason whatsoever when a better option exists

**Comment by F1F7Y** [comment]
> 

you can also `vector(x, y, z)` iirc

**Comment by GeckoEidechse** [comment]
Someone giving the suggested changes a final look and test maybe would be appreciated <3

**Comment by uniboi** [comment]
> Someone giving the suggested changes a final look and test maybe would be appreciated <3

What changes are you referring to? I thought only formatting has been changed so the test mod I sent earlier should still work fine?

**Comment by GeckoEidechse** [comment]
> > Someone giving the suggested changes a final look and test maybe would be appreciated <3
> 
> What changes are you referring to? I thought only formatting has been changed so the test mod I sent earlier should still work fine?

The commits that were added since the last review basically. I think it was just formatting as you said (don't have the setup rn to check). For testing, just running newest version of PR with your mod again should be enough. It's just to make sure there weren't any accidental errors syntax or whatever added since last change. Usually I'd do this myself but I'm on holiday atm :P

**Comment by GeckoEidechse** [comment]
> image @GeckoEidechse everything works now and images are fixed. This is imo ready to merge

Aight, Imma put my trust in you :P

Time to merge :D

**Comment by GeckoEidechse** [comment]
Welp, it's been fun not having to cherry-pick commits for patch releases ._.

**Comment by x3Karma** [comment]
WOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO

**Pull request merged**